Log reader tool

JeffreyDevloo opened the pull request:
Introduces a log file parser.

Can be utitlized as
```
from ovs.log.log_reader import LogFileTimeParser
search_locations = ['/var/log/upstart/ovs-volumedriver_geo-accelerated-alba.log']
perf_nodes = ['172.20.20.31','172.20.20.32','172.20.20.33','172.20.20.34','172.20.20.131','172.20.20.231']
ctrl_nodes = ['172.20.20.102', '172.20.20.201', '172.20.20.1']
all_nodes = [] + perf_nodes + ctrl_nodes
LogFileTimeParser.execute_search_on_remote(start='2016-09-01 13:46:00', end='2016-09-01 13:50:00', search_locations=files, hosts=all_nodes, suppress_return=True, password=None)
```
Which will collect the logs between the specified timestamps for all specified nodes and move the collected output to /tmp/file_reader_cache_all

Works on both 14.04 and 16.04.

JeffreyDevloo committed:
A tool to read logs on multiple nodes
Can search for error patterns, searches between timestamps
Future features will be
- supporting journalctl for ubuntu16.04
- Catching streamoutput from certain processes
- Reading in rotated logfiles

JeffreyDevloo committed:
Merge branch 'master' into log_reader_tool

JeffreyDevloo committed:
Added start of ubuntu version differences

JeffreyDevloo committed:
Merge branch 'master' into log_reader_tool

JeffreyDevloo committed:
Merge branch 'master' into log_reader_tool

JeffreyDevloo committed:
Reads journalctl for service info in 16.04

JeffreyDevloo committed:
Now fully static

JeffreyDevloo committed:
Removed debug statement

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+import time
+import platform
+import subprocess
+from stat import *
```
```import stat``` and then ```stat.foo```

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+class LogFileTimeParser(object):
+    """
+    # @TODO Support for rotated log files
+    # @TODO Support Ubuntu 16.04 journalctl
```
Not yet implemented?

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+
+    Common usage: validate log files during testing
+    """
+    version = "0.02"
```
Use ```'``` instead of ```"```.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+    LIMIT = 75  # arbitrary, allow for a VERY large file, but stop it if it runs away
+    PATHS_TO_FILES = ['/var/log/ovs/lib.log', '/var/log/ovs/extensions.log']
+
+    INTERNAL_MAPPING = {
```
It suddenly looks like javascript.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+    }
+
+    # Timeout for ssh: 0.0 means no timeout
+    TIME_OUT = 0.0
```
```TIMEOUT``` is one word

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+        return execution_mode
+
+    @staticmethod
+    def _execute_command(command, wait=True, shell=True):
```
Can we use the ```SSHClient``` for that?

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+    @staticmethod
+    def _check_available_space(hosts, username, password):
+        required_space = 0
+        output_dir = "/".join(LogFileTimeParser.FILE_PATH_REMOTE.strip("/").split('/')[:1])
```
Like ```LogFileTimeParser.FILE_PATH_REMOTE.strip('/').rsplit('/', 1)[0]```?

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+        from ovs.extensions.generic.sshclient import SSHClient
+        for host in hosts:
+            ssh_client = SSHClient(host, username=username, password=password)
+            for f in LogFileTimeParser.PATHS_TO_FILES:
```
Please use some more letters than just ```f```.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+        if not start:
+            start = (datetime.today() - timedelta(hours=1)).strftime("%Y-%m-%d %H:%M:%S")
+        if not end:
+            end = datetime.today().strftime("%Y-%m-%d %H:%M:%S")
```
This might result in an unexpected window if executed right before the day ends.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+
+        if search_locations is None:
+            execution_mode = LogFileTimeParser._get_exection_mode()
+            # search_locations = LogFileTimeParser.PATHS_TO_FILES
```
That's probably an obsolete comment

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+                search_locations.append(info.get(execution_mode))
+
+        if not hosts:
+            hosts = [LogFileTimeParser._clean_text(sr.ip) for sr in StorageRouterList.get_storagerouters()]
```
Is that ```_clean_text``` required for an ip address?

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+                elif mode == 'error-search':
+                    # Execute search
+                    results = remotes[host].LogFileTimeParser.search_for_errors(start=start, end=end,
+                                                                                  paths_to_file=search_locations,
```
Misalignment and inconsistent argument names.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+                                                                                  python_error=python_error)
+                # Append output to cache
+                with open(LogFileTimeParser.FILE_PATH_REMOTE, 'a') as f2:
+                    f2.write(str(results))
```
Again, slightly more meaningful names. We might get confused here.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+    @staticmethod
+    def _parse_date(text, validate=True):
+        """
+
```
Remove whitespace, or add something interesting.

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+            line = opened_file.readline(buf_size)
+        except:
+            raise IOError("File I/O Error")
+        if line == '':
```
What about empty lines in the logfiles?

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -0,0 +1,527 @@
+                cleaned = cleaned.replace(old, new)
+            return cleaned
+        except UnicodeDecodeError:
+            print 'UnicodeDecodeError with output: {0}'.format(text)
```
For a better algorithm, please check the ```SSHClient```.

JeffreyDevloo committed:
Requested changes

JeffreyDevloo committed:
More changes

khenderick commented on ovs/log/log_reader.py:
```diff
@@ -89,33 +77,17 @@ def _get_exection_mode():
-            for f in LogFileTimeParser.PATHS_TO_FILES:
-                command = "ls -al {0} | cut -d ' ' -f 5".format(f)
+            for file_path in LogFileTimeParser.PATHS_TO_FILES:
+                command = 'ls -al {0} | cut -d ' ' -f 5'.format(file_path)
```
This won't work

JeffreyDevloo committed:
16.04 supports both files and journal
Now exists faster if the date range is not in the file
All default values get assigned by a function now
